Playing video in Safari can generate non-monotonic 'timeupdate' events

jyavenard opened the pull request:
#### 0e4cfc6b8f5fa42166bb9e05a7e4ed227ec91f79
<pre>
Playing video in Safari can generate non-monotonic &apos;timeupdate&apos; events
<a href="https://bugs.webkit.org/show_bug.cgi?id=285656">https://bugs.webkit.org/show_bug.cgi?id=285656</a>
<a href="https://rdar.apple.com/142275184">rdar://142275184</a>

Reviewed by Eric Carlson.

When playback is about to start, a rateChanged notification is being sent
by the MediaPlayer running in the GPU process, followed shortly by currentTimeChanged
once playback has actually started.
The MediaPlayerPrivateRemote started to estimate the time progress upon
receiving the rateChanged notification only to go back to zero once again
following the currentTimeChanged update ; making the currentTime value go
backward.
We now stop estimating the currentTime progress when we get notified of
a rate change, returning the cached time instead until the next currentTimeChanged.

While testing, an additional issue was exposed, where currentTime could temporarily
exceed the video&apos;s duration, only to go back to be exactly the duration
once playback actually ended.
The HTMLMediaElement was also estimating the currentTime progress in an attempt
to reduce the number of calls to MediaPlayer::currentTime().
This estimation was occurring whenever the MediaPlayerPrivate::maximumDurationToCacheMediaTime
returned a value greater than 0, the default value being 0 (so no estimate was to occur
and the HTMLMediaElement was to always query the MediaPlayer&apos;s currentTime)
The only MediaPlayerPrivate overriding maximumDurationToCacheMediaTime
was the MediaPlayerPrivateRemote which set a value of 0.2s.
The MediaPlayerPrivateRemote is already estimating the currentTime, there&apos;s
as such no benefit for the HTMLMediaElement to cache the time and also
estimate the progress.
Due to the above, the time estimation found in the HTMLMediaElement was dead
code until the MediaPlayerPrivateRemote was introduced, only to be revived
unnecessarily.
It could also incorrectly report the time being greater than the video&apos;s element
if the currentTime was ever read after (duration-maximumDurationToCacheMediaTime).

We can remove the handling of cache time in the HTMLMediaElement as it introduced
error and was redundant over the MediaPlayerPrivateRemote&apos;s own estimation.

Added a test.

* LayoutTests/media/video-currenttime-monotonic-expected.txt: Added.
* LayoutTests/media/video-currenttime-monotonic.html: Added.
* Source/WebCore/WebCore.order:
* Source/WebCore/html/HTMLMediaElement.cpp: Remove all code related to estimating the currentTime from the last time m_cachedTime had been updated.
(WebCore::HTMLMediaElement::prepareForLoad):
(WebCore::HTMLMediaElement::mediaPlayerReloadAndResumePlaybackIfNeeded):
(WebCore::HTMLMediaElement::setReadyState):
(WebCore::HTMLMediaElement::fastSeek):
(WebCore::HTMLMediaElement::seekWithTolerance):
(WebCore::HTMLMediaElement::clearSeeking):
(WebCore::HTMLMediaElement::invalidateOfficialPlaybackPosition):
(WebCore::HTMLMediaElement::currentMediaTime const):
(WebCore::HTMLMediaElement::setPlaybackRate):
(WebCore::HTMLMediaElement::playInternal):
(WebCore::HTMLMediaElement::mediaPlayerTimeChanged):
(WebCore::HTMLMediaElement::mediaPlayerRateChanged):
(WebCore::HTMLMediaElement::updatePlayState):
(WebCore::HTMLMediaElement::refreshCachedTime const): Deleted.
(WebCore::HTMLMediaElement::invalidateCachedTime const): Deleted.
* Source/WebCore/html/HTMLMediaElement.h:
* Source/WebCore/platform/graphics/MediaPlayer.cpp:
(WebCore::MediaPlayer::maximumDurationToCacheMediaTime const): Deleted.
* Source/WebCore/platform/graphics/MediaPlayer.h:
* Source/WebCore/platform/graphics/MediaPlayerPrivate.h:
(WebCore::MediaPlayerPrivateInterface::maximumDurationToCacheMediaTime const): Deleted.
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.cpp:
(WebKit::RemoteMediaPlayerProxy::getConfiguration):
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
(WebKit::MediaPlayerPrivateRemote::TimeProgressEstimator::currentTimeWithLockHeld const):
Ensure to only return a new time value if it has actually progressed since the last call.
Small time drifts can occur between the GPU and WebContent process.
(WebKit::MediaPlayerPrivateRemote::TimeProgressEstimator::setTime):
(WebKit::MediaPlayerPrivateRemote::seeked):
(WebKit::MediaPlayerPrivateRemote::timeChanged):
(WebKit::MediaPlayerPrivateRemote::rateChanged):
(WebKit::MediaPlayerPrivateRemote::playbackStateChanged):
(WebKit::MediaPlayerPrivateRemote::currentTimeChanged):
(WebKit::MediaPlayerPrivateRemote::maximumDurationToCacheMediaTime const): Deleted.
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.h:
* Source/WebKit/WebProcess/GPU/media/RemoteMediaPlayerConfiguration.h:
* Source/WebKit/WebProcess/GPU/media/RemoteMediaPlayerConfiguration.serialization.in:

Canonical link: <a href="https://commits.webkit.org/288722@main">https://commits.webkit.org/288722@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5a74ffb3cd9412da39cb5f0e1cdffd0186336b51

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/84160 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/3781 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/38464 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/89236 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/35169 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/86245 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/3869 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/11752 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/65456 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/23294 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/87206 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/2897 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/76478 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/45749 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/2842 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/30707 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/34217 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/73774 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/31470 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/90616 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/11427 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/8290 "Found 2 new test failures: fast/selectors/selection-window-inactive-stroke-color.html imported/w3c/web-platform-tests/webcodecs/audio-encoder.https.any.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/73910 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/11653 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/72303 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/73114 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18101 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/17425 "Passed tests") | [⏳ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/macOS-Ventura-Release-WK2-Intel-Tests-EWS "Waiting to run tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/2772 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/11379 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/16855 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/11228 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/14704 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/13001 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

eric-carlson commented on Source/WebCore/html/HTMLMediaElement.cpp:
```diff
@@ -3987,34 +3980,9 @@ void HTMLMediaElement::setSeeking(bool seeking)
-    m_clockTimeAtLastCachedTimeUpdate = MonotonicTime::now();
-}
-
 void HTMLMediaElement::invalidateCachedTime() const
```
As we discussed, I think you should remove the _all_ of the caching code from HTMLMediaElement

jyavenard committed:
Playing video in Safari can generate non-monotonic 'timeupdate' events
https://bugs.webkit.org/show_bug.cgi?id=285656
rdar://142275184

Reviewed by Eric Carlson.

When playback is about to start, a rateChanged notification is being sent
by the MediaPlayer running in the GPU process, followed shortly by currentTimeChanged
once playback has actually started.
The MediaPlayerPrivateRemote started to estimate the time progress upon
receiving the rateChanged notification only to go back to zero once again
following the currentTimeChanged update ; making the currentTime value go
backward.
We now stop estimating the currentTime progress when we get notified of
a rate change, returning the cached time instead until the next currentTimeChanged.

While testing, an additional issue was exposed, where currentTime could temporarily
exceed the video's duration, only to go back to be exactly the duration
once playback actually ended.
The HTMLMediaElement was also estimating the currentTime progress in an attempt
to reduce the number of calls to MediaPlayer::currentTime().
This estimation was occurring whenever the MediaPlayerPrivate::maximumDurationToCacheMediaTime
returned a value greater than 0, the default value being 0 (so no estimate was to occur
and the HTMLMediaElement was to always query the MediaPlayer's currentTime)
The only MediaPlayerPrivate overriding maximumDurationToCacheMediaTime
was the MediaPlayerPrivateRemote which set a value of 0.2s.
The MediaPlayerPrivateRemote is already estimating the currentTime, there's
as such no benefit for the HTMLMediaElement to cache the time and also
estimate the progress.
Due to the above, the time estimation found in the HTMLMediaElement was dead
code until the MediaPlayerPrivateRemote was introduced, only to be revived
unnecessarily.
It could also incorrectly report the time being greater than the video's element
if the currentTime was ever read after (duration-maximumDurationToCacheMediaTime).

We can remove the handling of cache time in the HTMLMediaElement as it introduced
error and was redundant over the MediaPlayerPrivateRemote's own estimation.

Added a test.

* LayoutTests/media/video-currenttime-monotonic-expected.txt: Added.
* LayoutTests/media/video-currenttime-monotonic.html: Added.
* Source/WebCore/WebCore.order:
* Source/WebCore/html/HTMLMediaElement.cpp: Remove all code related to estimating the currentTime from the last time m_cachedTime had been updated.
(WebCore::HTMLMediaElement::prepareForLoad):
(WebCore::HTMLMediaElement::mediaPlayerReloadAndResumePlaybackIfNeeded):
(WebCore::HTMLMediaElement::setReadyState):
(WebCore::HTMLMediaElement::fastSeek):
(WebCore::HTMLMediaElement::seekWithTolerance):
(WebCore::HTMLMediaElement::clearSeeking):
(WebCore::HTMLMediaElement::invalidateOfficialPlaybackPosition):
(WebCore::HTMLMediaElement::currentMediaTime const):
(WebCore::HTMLMediaElement::setPlaybackRate):
(WebCore::HTMLMediaElement::playInternal):
(WebCore::HTMLMediaElement::mediaPlayerTimeChanged):
(WebCore::HTMLMediaElement::mediaPlayerRateChanged):
(WebCore::HTMLMediaElement::updatePlayState):
(WebCore::HTMLMediaElement::refreshCachedTime const): Deleted.
(WebCore::HTMLMediaElement::invalidateCachedTime const): Deleted.
* Source/WebCore/html/HTMLMediaElement.h:
* Source/WebCore/platform/graphics/MediaPlayer.cpp:
(WebCore::MediaPlayer::maximumDurationToCacheMediaTime const): Deleted.
* Source/WebCore/platform/graphics/MediaPlayer.h:
* Source/WebCore/platform/graphics/MediaPlayerPrivate.h:
(WebCore::MediaPlayerPrivateInterface::maximumDurationToCacheMediaTime const): Deleted.
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.cpp:
(WebKit::RemoteMediaPlayerProxy::getConfiguration):
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
(WebKit::MediaPlayerPrivateRemote::TimeProgressEstimator::currentTimeWithLockHeld const):
Ensure to only return a new time value if it has actually progressed since the last call.
Small time drifts can occur between the GPU and WebContent process.
(WebKit::MediaPlayerPrivateRemote::TimeProgressEstimator::setTime):
(WebKit::MediaPlayerPrivateRemote::seeked):
(WebKit::MediaPlayerPrivateRemote::timeChanged):
(WebKit::MediaPlayerPrivateRemote::rateChanged):
(WebKit::MediaPlayerPrivateRemote::playbackStateChanged):
(WebKit::MediaPlayerPrivateRemote::currentTimeChanged):
(WebKit::MediaPlayerPrivateRemote::maximumDurationToCacheMediaTime const): Deleted.
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.h:
* Source/WebKit/WebProcess/GPU/media/RemoteMediaPlayerConfiguration.h:
* Source/WebKit/WebProcess/GPU/media/RemoteMediaPlayerConfiguration.serialization.in:

Canonical link: https://commits.webkit.org/288722@main
